### PR TITLE
update go-nix hash and refactor index.go to satisfy the new go-nix

### LINKED ISSUE #7: Add S3 backend
Some people's binary cache live in S3.

This will require to extend the go-nix library so that it can parse a store URI and return the right backend for us.

### PR DESCRIPTION
**Pull request by Rizary**
fixes #7

### CODE REVIEW

File: api/unpack/index.go [hunk 1]
```diff
@@ -177,8 +182,8 @@ func getEnv(name, def string) string {
 // TODO: consider keeping a LRU cache
 func getNarInfo(key string) (*libstore.NarInfo, error) {
 	path := fmt.Sprintf("%s.narinfo", key)
-	fmt.Println("Fetching the narinfo:", path, "from:", nixCache.CacheURI)
-	r, err := nixCache.GetFile(path)
+	fmt.Println("Fetching the narinfo:", path, "from:", nixCache.URL)
```
Comment (zimbatm) [inline api/unpack/index.go:185]:
```suggestion
	fmt.Println("Fetching the narinfo:", path, "from:", nixCache.URL())
```
URL is now a function

File: api/unpack/index.go [hunk 2]
```diff
@@ -18,9 +19,13 @@ import (
...
+var ctx = context.Background()
+
+var nixCache, err = libstore.NewBinaryCacheReader(ctx, "s3://nar?region=eu-west-1&endpoint=http://127.0.0.1:9000&scheme=http")
+
+// libstore.HTTPBinaryCacheStore{
+// 	CacheURI: getEnv("NAR_CACHE_URI", "https://cache.nixos.org"),
+// }
 
 // Handler is the entry-point for @now/go as well as the stub main.go net/http
 func Handler(w http.ResponseWriter, req *http.Request) {
```
Comment (zimbatm) [inline api/unpack/index.go:31]: Use the context of the request instead:
```suggestion
func Handler(w http.ResponseWriter, req *http.Request) {
  ctx := req.Context()
```

File: api/unpack/index.go [hunk 3]
```diff
@@ -18,9 +19,13 @@ import (
 // MountPath is where this handler is supposed to be mounted
 const MountPath = "/nix/store/"
 
-var nixCache = libstore.HTTPBinaryCacheStore{
-	CacheURI: getEnv("NAR_CACHE_URI", "https://cache.nixos.org"),
-}
+var ctx = context.Background()
```
Comment (zimbatm) [inline api/unpack/index.go:22]: Context objects are supposed to be passed around so they can be controlled. So setting one top-level here defeats its purpose.

File: api/unpack/index.go [hunk 4]
```diff
@@ -177,8 +182,8 @@ func getEnv(name, def string) string {
 // TODO: consider keeping a LRU cache
 func getNarInfo(key string) (*libstore.NarInfo, error) {
```
Comment (zimbatm) [inline api/unpack/index.go:183]:
```suggestion
func getNarInfo(key string) (ctx context.Context, *libstore.NarInfo, error) {
```

File: api/unpack/index.go (hunk 3 above)
Comment (Rizary) [reply api/unpack/index.go:22]: I agree, but I have to pass `ctx` to this function
```
var nixCache, err = libstore.NewBinaryCacheReader(ctx, "s3://nar?region=eu-west-1&endpoint=http://127.0.0.1:9000&scheme=http")
```

Do you have any suggestion?

Comment (zimbatm) [reply api/unpack/index.go:22]: Did you try passing `context.Background()` directly to that function?

Comment (zimbatm) [reply api/unpack/index.go:22]: That module design is not super clean as it was influenced by the restrictions from Vercel.

A better way would be to create a new struct that obeys the http.Handler interface, and initialize all of that is the top-level main function. That would make it more composable and more easily testable.

Comment (zimbatm) [reply api/unpack/index.go]: pushed fix in 477d80549859a47897d4a2dceb6d9e13d1a1ce9b

**Pull request merged**